apisix/plugins/sls-logger.lua

```diff
@@ -0,0 +1,213 @@
+    if not ok then
+        err_msg = "failed to connect to TCP server: host[" .. route_conf.host
+        .. "] port[" .. tostring(route_conf.port) .. "] err: " .. err
+        core.log.error(err_msg)
+        return false, err_msg
+    end
+
+    ok, err = sock:sslhandshake(true, nil, false)
+    if not ok then
+        err_msg = "failed to to perform TLS handshake to TCP server: host["
+        .. route_conf.host .. "] port[" .. tostring(route_conf.port) .. "] err: " .. err
+        core.log.error(err_msg)
+        return false, err_msg
+    end
+
+    core.log.info("sls logger send data ", log_message)
+    ok, err = sock:send(log_message)
+    if not ok then
+        res = false
+        can_close = true
+        err_msg = "failed to send data to TCP server: host[" .. route_conf.host
+                  .. "] port[" .. tostring(route_conf.port) .. "] err: " .. err
+        core.log.error(err_msg)
+    else
+        ok, err = sock:setkeepalive(120 * 1000, 20)
+        if not ok then
+            can_close = true
+            core.log.error("failed to set socket keepalive: host[" .. route_conf.host
+            .. "] port[" .. tostring(route_conf.port) .. "] err: " .. err)
+        end
+    end
+
+    if  can_close then
+        ok, err = sock:close()
+        if not ok then
+            core.log.error("failed to close the TCP connection, host[",
+            route_conf.host, "] port[", route_conf.port, "] ", err)
+        end
+    end
+
+    return res, err_msg
+end
+
+-- remove stale objects from the memory after timer expires
+local function remove_stale_objects(premature)
+    if premature then
+        return
+    end
+
+    for key, batch in ipairs(buffers) do
+        if #batch.entry_buffer.entries == 0 and #batch.batch_to_process == 0 then
+            core.log.warn("removing batch processor stale object, route id:", tostring(key))
+            buffers[key] = nil
+        end
+    end
+
+    stale_timer_running = false
+end
+
+local function combine_syslog(entries)
+    local data
+    for _, entry in ipairs(entries) do
+        if not data then
+           data = entry.data
+        end
+
+        data = data .. entry.data
+        core.log.info(entry.data)
+    end
+
+    return data
+end
+
+local function handle_log(entries)
+    local data = combine_syslog(entries)
+    if not data then
+        return true
+    end
+
+    -- get the config from entries, replace of local value
+    return send_tcp_data(entries[1].route_conf, data)
+end
+
+-- log phase in APISIX
+function _M.log(conf)
+    local entry = log_util.get_full_log(ngx, conf)
+    if not entry.route_id then
+        core.log.error("failed to obtain the route id for sys logger")
+        return
+    end
+
+    local json_str, err = core.json.encode(entry)
+    if not json_str then
+        core.log.error('error occurred while encoding the data: ', err)
+        return
+    end
+
+    local rf5424_data = rf5424.encode("SYSLOG", "INFO", ngx.var.host
```

Comment:
in APISIX, we shuold use `ctx.var.****` to fetch nginx variable.

here is an example: https://github.com/apache/apisix/blob/master/apisix/plugins/limit-count.lua#L166

Reply:
fix